### ISSUE TITLE
Crash on Keychron Q3 HE

### ISSUE THREAD

**Issue by King-of-Infinite-Space** [calamity-inc/universal-analog-plugin#6]
Hi, thanks for developing this plugin. I followed the readme to install Wooting SDK and this plugin, and flashed custom firmware on Keychron Q3 HE. When I test it on https://github.com/WootingKb/wooting-analog-midi, the program can see the keyboard as a connected device, but freezes and crashes (sometimes on start, sometimes after pressing a few keys). Is this supposed to work? Or is there another way to test midi functionality?

**Comment by Sainan** [comment]
It's supposed to work, yeah. Is your Wooting Analog SDK up-to-date?

**Comment by Sainan** [comment]
Also works fine on my machine.

![image](https://github.com/user-attachments/assets/6b778b80-4ee4-4de0-9eaf-2547d5f082fd)

**Comment by Sainan** [comment]
It might help if you clone wooting-analog-midi, and run it manually:
```batch
SET NODE_OPTIONS=--openssl-legacy-provider
yarn tauri dev
```
If there's any error, it will show it in the console.

**Comment by King-of-Infinite-Space** [comment]
1. Yes, I have the Wooting Analog SDK v0.8.0 using the msi installer from the latest release.
2. This is my folder structure
	```
	WootingAnalogPlugins
	├── universal-analog-plugin
	│   ├── abiv0.dll
	│   └── abiv1.dll
	├── wooting-analog-plugin
	│   └── wooting_analog_plugin.dll
	└── wooting-test-plugin
	    └── wooting_analog_test_plugin.dll
	```
	I also tried using only the `universal-analog-plugin-with-wooting-device-support` like the screenshot but same issue

3. I ran the dev server. I only get this message when the app starts. No log when it crashes.
	```
	Device 0 is DeviceInfo { vendor_id: 13364, product_id: 2864, manufacturer_name: "Keychron", device_name: "Keychron Q3 HE", device_id: 875825968, device_type: Keyboard }
	```

**Comment by Sainan** [comment]
Well, I'm also using the stock SDK v0.8.0 and otherwise your setup looks fine.
- Maybe you need the Visual C++ redistributables if your system is fresh-ish: https://aka.ms/vs/17/release/vc_redist.x64.exe
- Alternatively you could try grabbing [a Soup build](https://github.com/calamity-inc/Soup/actions/runs/12328117253/artifacts/2320739508), extracting the soup.exe from the CLI folder, and running "soup.exe keyboard" in a console window. This would show if the underlying soup::AnalogueKeyboard functionality is working on your system.

**Comment by King-of-Infinite-Space** [comment]
`soup.exe keyboard` works. Looks like it's based on HID? Maybe I can implement my own solution for MIDI.

I don't know Rust but I asked LLM to write [an MVP](https://github.com/King-of-Infinite-Space/wooting-analog-demo/blob/main/src/main.rs) based on `wooting-analog-midi-core\src\lib.rs` to print output from the SDK. I get the same crashing behavior.
```
     Running `target\debug\wooting-analog-demo.exe`
10:06:50 [INFO] Starting Wooting Analog SDK!
10:06:50 [INFO] Analog SDK Successfully initialised with 1 devices
Device 0: VID: 3434, PID: 0B30, Manufacturer: Keychron, Name: Keychron Q3 HE, ID: 875825968, Type: Keyboard
Press any key on the Wooting keyboard to see the analog value. Press Ctrl+C to exit.
Key ID: 30, Analog Value: 0.9447
Key ID: 30, Analog Value: 0.4043
Key ID: 30, Analog Value: 0.0000
error: process didn't exit successfully: `target\debug\wooting-analog-demo.exe` (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION)
```

**Comment by Sainan** [comment]
ACCESS_VIOLATION is certainly interesting. I guess I can't exclude the possibility that there's some issue in my code that I just can't reproduce on my machine.

I've compiled a build with logging and a PDB attached, this might help narrow it down: [universal-analog-plugin.zip](https://github.com/user-attachments/files/18140038/universal-analog-plugin.zip)

**Comment by King-of-Infinite-Space** [comment]
Not sure if I'm doing it right - using the new dll and running my demo via the debugger in Visual Studio

<img width="1280" alt="screenshot" src="https://github.com/user-attachments/assets/6f647682-7df2-43d4-a081-3b18370c1b00" />

**Comment by Sainan** [comment]
Well, that doesn't say much. What about the console logs?

**Comment by King-of-Infinite-Space** [comment]
```
> $env:RUST_LOG = "debug"
> .\demo\target\debug\wooting-analog-demo.exe
09:29:34 [INFO] Starting Wooting Analog SDK!
[2024-12-17T09:29:34Z INFO  wooting_analog_sdk::sdk] No plugins found in "C:\\Program Files\\WootingAnalogPlugins"
[2024-12-17T09:29:34Z INFO  wooting_analog_sdk::sdk] Loading plugin: "C:\Program Files\WootingAnalogPlugins\universal-analog-plugin\abiv1-pluswooting.dll"
[2024-12-17T09:29:34Z WARN  wooting_analog_sdk::sdk] Unable to determine the Plugin's SemVer!
[2024-12-17T09:29:34Z WARN  wooting_analog_sdk::sdk] Failed to find constructor (_plugin_create symbol): GetProcAddress failed: 找不到指定的程序。(translation: The specified procedure could not be found.) (os error 127)
[2024-12-17T09:29:34Z INFO  wooting_analog_sdk::sdk] Didn't find _plugin_create, assuming it's a C plugin
[2024-12-17T09:29:34Z INFO  wooting_analog_sdk::cplugin] Got cplugin abi: 1
[2024-12-17T09:29:34Z INFO  wooting_analog_sdk::sdk] Loaded plugin: "Universal Analog Plugin"
[2024-12-17T09:29:34Z DEBUG wooting_analog_sdk::sdk] Loaded 1 plugins from "C:\\Program Files\\WootingAnalogPlugins\\universal-analog-plugin"
New device: Keychron Q3 HE
Discovered 1 initial devices
- Keychron Q3 HE
[2024-12-17T09:29:34Z DEBUG wooting_analog_sdk::sdk] SDKResult(Ok(1))
[2024-12-17T09:29:34Z INFO  wooting_analog_sdk::sdk] 1 plugins successfully initialised
09:29:34 [INFO] Analog SDK Successfully initialised with 1 devices
Device 0: VID: 3434, PID: 0B30, Manufacturer: Keychron, Name: Keychron Q3 HE, ID: 875825968, Type: Keyboard
Press any key on the Wooting keyboard to see the analog value. Press Ctrl+C to exit.
Key ID: 9, Analog Value: 0.0596
Key ID: 9, Analog Value: 0.2511
Key ID: 9, Analog Value: 1.0000
Key ID: 9, Analog Value: 0.9702
Key ID: 9, Analog Value: 0.1745
Key ID: 9, Analog Value: 0.0000
```

The program exits without error message. Sometimes it exits before `SDKResult(Ok(1))`.

**Comment by Sainan** [comment]
Hmm, well that's not really telling me anything, damn. Are you on Windows 10 or 11?

**Comment by King-of-Infinite-Space** [comment]
Win11 24H2 (26100.2605)

**Comment by Sainan** [comment]
Well, I personally use Windows 10. I do have an ARM laptop with Windows 11, so I installed wooting-analog-midi, wooting-analog-sdk, & universal-analog-plugin on it, plugged in my Q3 HE, and it also worked fine there.

**Comment by King-of-Infinite-Space** [comment]
Also seems to be working on my other computer running Win11 (23H2)...

**Comment by King-of-Infinite-Space** [comment]
I guess I'll use the midi functionality on the other computer for now. Thanks for helping track it down.

**Comment by Sainan** [comment]
I'm unsure about the exact Windows 11 version on my laptop since it constantly updates itself. As I just booted it up again, I was greeted with the Copilot app having been added to the taskbar, but it claims to still be 23H2. So, maybe there's an issue with 24H2 but unless it gets rolled out further, I'm not sure I'll be able to repro it.

**Comment by Azarattum** [comment]
I was able to reproduce the crash and even got the [dump from it](https://github.com/user-attachments/files/19262382/crash.zip) (filter by `usb.src == 8.3.2 || usb.dst == 8.3.2`). It seems that the host asks for a version for some reason (with `a9 01` packet) in the middle of polling. When the keyboard responds with its version it seems to be that the client was expecting the key matrix packet instead. Hence the access violation and the crash. @Sainan, any idea why the version packet would be sent in the first place?

![Image](https://github.com/user-attachments/assets/9468d4ae-11b0-4636-93ab-5a67be43f75e)

**Comment by Sainan** [comment]
I don't get why that would cause a crash. All packets are 32 bytes, so at worst it would just read some garbage data.

**Comment by Azarattum** [comment]
Wouldn't reading garbage data cause a crash? You operate on this data later, so it would almost be undefined behavior, no? Also, if my understanding goes correctly, doing so
```c
hid.sendReport(...);
... = hid.receiveReport();
```
is pretty unreliable, since whatever may come from the second report given the concurrent requests.

The version call is probably just the app periodically scanning for new devices, therefore calling `AnalogueKeyboard::getAll` which checks for the version every time.

**Comment by Sainan** [comment]
Yeah, I figured that as well so added mutexing code: https://github.com/calamity-inc/Soup/commit/ead6f32854c23bd844b48b8ea85078e2a31941a0

But just garbage data alone is not an issue, especially as it's ensured that to be in-bounds anyway.

**Comment by Azarattum** [comment]
Great, I'll try with the latest commit now. I've also just confirmed that just doing nothing (not touching any keys) while the midi app is open also causes the crash after some time. So it must be a periodic device check.

**Comment by Azarattum** [comment]
I don't think it worked. The requests are still interleaved. Also I notice some fake key presses registered in the midi app (probably because of this version check). I think it was the "Home" button.

![Image](https://github.com/user-attachments/assets/089b6884-ab6a-4769-b596-7d40ab994443)

**Comment by Azarattum** [comment]
Btw, it asks for the version many times before it actually crashes. So there probably some another factor there. Some kind of a race condition?

**Comment by Azarattum** [comment]
Although it is consistent that for every crash second to last sent packet was the version check and the last was a key poll.

**Comment by Azarattum** [comment]
I added some logging. It creates new instances of `AnalogueKeyboard` like every 2 seconds. This doesn't seem right at all. Then it crashes after receiving polling packet at the version packet and then trying to poll:
```
sending version report...
got version back: A9 31 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
sending travel report...

** EOF, CRASH! **
```

This is probably because it switches mode from full analog report to a single key (since the version has no `45` indicator).

**Comment by Azarattum** [comment]
And it also seems like mutexes are not shared between these instances...

**Comment by Azarattum** [comment]
Btw, Cyberpunk also crashes with:
```
Error reason: Unhandled exception
Expression: EXCEPTION_ACCESS_VIOLATION (0xC0000005)
Message: The thread caused a user-mode data execution prevention (DEP) violation at 0x21.
File: <Unknown>(0)
```
before even launching the game. So this issue isn't specific to the midi app.

**Comment by Azarattum** [comment]
Should be fixed in https://github.com/calamity-inc/Soup/pull/145. @King-of-Infinite-Space, let us know if it works for you too. Here is a build with the fix (for Windows): [universal-analog-plugin.zip](https://github.com/user-attachments/files/19263745/universal-analog-plugin.zip)

**Comment by King-of-Infinite-Space** [comment]
I confirm it works. Thank you!
